### PR TITLE
[feature] kv transfer support of ascend npu

### PR DESCRIPTION
**Pull request by ping1jing2**
## Motivation

Support for kv transfer in the Ascend cluster in the PD disaggregation scenario.

## Modifications

`AscendTransferEngine` for kv transfer and can support both `mooncake` and `AscendTransferEngine`.

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/disaggregation/ascend/transfer_engine.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+    def __init__(self, hostname: str, ascend_url: str, npu_id: int, disaggregation_mode: DisaggregationMode,
+                 ib_device: str, ascend_mooncake: bool):
+        if not ascend_mooncake:
+            try:
+                from mf_adapter import TransferEngine
+            except ImportError as e:
+                raise ImportError(
+                    "Please install AscendTransferEngine, run SGLang with AscendTransferEngine"
+                ) from e
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/ascend/transfer_engine.py]: Maybe provide an instruction to tell the user how to install `mf_adapter`?

File: python/sglang/srt/disaggregation/ascend/transfer_engine.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                f"Unsupported DisaggregationMode: {disaggregation_mode}"
+            )
+        if not self.ascend_mooncake:
+            self.session_id = f"{self.hostname}:{self.engine.get_rpc_port()}"
+        self.initialize(
+            hostname=self.hostname,
+            device_name=self.ib_device
+        )
+        if self.ascend_mooncake:
+            self.session_id = f"{self.hostname}:{self.engine.get_rpc_port()}"
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/ascend/transfer_engine.py]: These lines seem confusing to me.

File: python/sglang/srt/disaggregation/ascend/transfer_engine.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+                self.role,
+                self.npu_id,
+            )
+        if ret_value != 0:
+            logger.error("Ascend Transfer Engine initialization failed.")
+            raise RuntimeError("Ascend Transfer Engine initialization failed.")
+
+    def batch_register(self, ptr, length):
+        try:
+            ret_value = self.engine.batch_register_memory(ptr, length)
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/ascend/transfer_engine.py]: Why does this batch operation not contain List input? I suppose the ptr should be `ptrs: List[int]`, and length should be `lengths: List[int]`? Or is this just a naming issue?

File: python/sglang/srt/disaggregation/ascend/transfer_engine.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        except Exception:
+            # Mark register as failed
+            ret_value = -1
+        if ret_value != 0:
+            logger.debug(f"Ascend memory  registration %s failed.", ptr)
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/ascend/transfer_engine.py]: Extra space in the log.

File: python/sglang/srt/disaggregation/ascend/transfer_engine.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+    def transfer_sync(
+        self, session_id: str, buffer: int, peer_buffer_address: int, length: int
+    ) -> int:
+        """Synchronously transfer data to the specified address."""
+        try:
+            # the first time: based on session_id (which contains remote_ip) to construct a queue pair, and cache the queue pair
+            # later: based on the cached queue pair to send data
+            torch_npu.npu.set_device(self.npu_id)
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/ascend/transfer_engine.py]: Should this line be placed in the `initialize` function? I don't see a reason why we need to repeat this process for each transfer op.

File: python/sglang/srt/disaggregation/ascend/transfer_engine.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    def batch_transfer_sync(
+        self,
+        session_id: str,
+        buffers: List[int],
+        peer_buffer_addresses: List[int],
+        lengths: List[int],
+    ) -> int:
+        """Synchronously transfer data to the specified addresses in batches."""
+        try:
+            torch_npu.npu.set_device(self.npu_id)
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/ascend/transfer_engine.py]: ditto

File: python/sglang/srt/disaggregation/base/conn.py [hunk 7]
```diff
@@ -20,8 +20,10 @@ class KVArgs:
     aux_data_ptrs: List[int]
     aux_data_lens: List[int]
     aux_item_lens: List[int]
+    ascend_mooncake: bool
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/base/conn.py]: I think this line is an ascend-specific config, it should not appear in the `KVArgs`. Maybe use an env var? Try `get_bool_env_var`.

File: python/sglang/srt/disaggregation/base/conn.py [hunk 8]
```diff
@@ -20,8 +20,10 @@ class KVArgs:
...
     ib_device: str
     ib_traffic_class: str
+    ascend_url: str
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/base/conn.py]: ditto

File: python/sglang/srt/disaggregation/decode.py [hunk 9]
```diff
@@ -210,6 +210,8 @@ def _init_kv_manager(self) -> BaseKVManager:
         )
 
         kv_args.ib_device = self.scheduler.server_args.disaggregation_ib_device
+        kv_args.ascend_url = self.scheduler.server_args.disaggregation_ascend_url
+        kv_args.ascend_mooncake = self.scheduler.server_args.ascend_mooncake
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/decode.py]: Ditto. Remove these lines and use an env var in the conn.py.

File: python/sglang/srt/disaggregation/prefill.py [hunk 10]
```diff
@@ -130,6 +130,8 @@ def _init_kv_manager(self) -> BaseKVManager:
             self.metadata_buffers.get_buf_infos()
         )
         kv_args.ib_device = self.scheduler.server_args.disaggregation_ib_device
+        kv_args.ascend_url = self.scheduler.server_args.disaggregation_ascend_url
+        kv_args.ascend_mooncake = self.scheduler.server_args.ascend_mooncake
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/prefill.py]: Ditto.

File: python/sglang/srt/disaggregation/utils.py [hunk 11]
```diff
@@ -94,8 +94,11 @@ def __init__(
         custom_mem_pool: torch.cuda.MemPool = None,
     ):
         self.custom_mem_pool = custom_mem_pool
-        device = "cuda" if self.custom_mem_pool else "cpu"
-
+        device = "cpu"
+        if is_npu():
+            device = "npu"
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/utils.py]: "cpu" is fine for normal usage. Is there a reason why you need this meta in the NPU? We need them in the gpu since we use custom_mem_pool to enable MNNVL, not sure about whether the `is_npu` logic here is required.

File: python/sglang/srt/managers/tokenizer_manager.py [hunk 12]
```diff
@@ -285,6 +286,17 @@ def __init__(
...
+                self.server_args.node_rank == 0
+                and is_npu()
+                and self.server_args.disaggregation_transfer_backend == "ascend"
+                and not self.server_args.ascend_mooncake
+            )
+            if is_create_store:
+                from mf_adapter import create_config_store
+
+                ascend_url = self.server_args.disaggregation_ascend_url
+                create_config_store(ascend_url)
```
Comment (ShangmingCai) [inline python/sglang/srt/managers/tokenizer_manager.py]: Do we need a try-except here? And ascend-specific configs should use env vars.

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 13]
```diff
@@ -574,32 +574,20 @@ def _create_buffers(self):
...
+                    self.size // self.page_size + 1,
+                    self.page_size,
+                    self.head_num,
+                    self.head_dim,
+                ),
+                dtype=self.store_dtype,
+                device=self.device,
+            )
+            self.k_buffer = self.kv_buffer[0]
+            self.v_buffer = self.kv_buffer[1]
```
Comment (ShangmingCai) [inline python/sglang/srt/mem_cache/memory_pool.py]: I am not sure about these changes. If it has no performance improvements, you should keep the original impl in case you break other features.

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 14]
```diff
@@ -922,18 +910,16 @@ def __init__(
...
+            self.kv_buffer = torch.zeros(
+                (
+                    layer_num,
+                    self.size // self.page_size + 1,
+                    self.page_size,
+                    self.kv_lora_rank + self.qk_rope_head_dim,
+                ),
+                dtype=self.store_dtype,
+                device=self.device,
+            )
```
Comment (ShangmingCai) [inline python/sglang/srt/mem_cache/memory_pool.py:998]: ditto

File: python/sglang/srt/server_args.py [hunk 15]
```diff
@@ -238,8 +238,10 @@ class ServerArgs:
     disaggregation_decode_dp: Optional[int] = None
     disaggregation_prefill_pp: Optional[int] = 1
     disaggregation_ib_device: Optional[str] = None
+    disaggregation_ascend_url: Optional[str] = None
```
Comment (ShangmingCai) [inline python/sglang/srt/server_args.py]: ditto. use env var.

File: python/sglang/srt/server_args.py [hunk 16]
```diff
@@ -238,8 +238,10 @@ class ServerArgs:
...
     num_reserved_decode_tokens: int = 512  # used for decode kv cache offload in PD
     pdlb_url: Optional[str] = None
+    ascend_mooncake: bool = False
```
Comment (ShangmingCai) [inline python/sglang/srt/server_args.py]: ditto.

File: python/sglang/srt/server_args.py [hunk 17]
```diff
@@ -1634,6 +1636,17 @@ def add_cli_args(parser: argparse.ArgumentParser):
...
+            "--disaggregation-ascend-url",
+            type=str,
+            default=ServerArgs.disaggregation_ascend_url,
+            help="(e.g., --disaggregation-ascend-url tcp://127.0.0.1:33233)",
+        )
+        parser.add_argument(
+            "--ascend-mooncake",
+            action="store_true",
+            help="Ascend Transfer mooncake",
+        )
```
Comment (ShangmingCai) [inline python/sglang/srt/server_args.py]: ditto.

File: python/sglang/srt/disaggregation/mooncake/conn.py [hunk 18]
```diff
@@ -234,6 +237,20 @@ def _connect(self, endpoint: str):
...
         return socket
 
+    def calculate_kv_addr(
+        self,
+        src_ptr: int,
+        dst_ptr: int,
+        prefill_index: list[int],
+        decode_index: list[int],
+        item_len: int,
+        data_len: int,
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/mooncake/conn.py]: `data_len` is only needed when registering. I am ok with the `calculate_kv_addr`, but I don't see these is a need to make these changes a util function since this calculation only appears once. It would be better to only keep the `init_engine` changes and revert all other changes in mooncake/conn.py, since this newly changed code is actually doing nothing.

**Review by ShangmingCai — COMMENTED** [review]
IIUC, most changes are related to the Ascend backend, which seems to be a placeholder right now. If you want to use current mooncake impl with ascend, I think you only need to change the `initialize` func of _mooncake/transfer_engine.py_, something like:
```
        if get_bool_env_var("SGLANG_MOONCAKE_WITH_ASCEND", "false"):
            hostname += f":{get_free_port()}:npu_{self.npu_id}"
            ret_value = self.engine.initialize(
                hostname,
                "P2PHANDSHAKE",
                "ascend",
                device_name if device_name is not None else "",
            )
```
Is it correct?

I think most changes are related to a new backend that hasn't seemed to be ready yet. A better choice is to open a new PR to support current mooncake with ascend backend first. Then we revisit this PR when _ascend/conn.py_ is ready.

File: python/sglang/srt/managers/tokenizer_manager.py [hunk 19]
```diff
@@ -285,6 +285,20 @@ def __init__(
             self.bootstrap_server = kv_bootstrap_server_class(
                 self.server_args.disaggregation_bootstrap_port
             )
+            is_create_store = (
+                self.server_args.node_rank == 0
+                and self.server_args.disaggregation_transfer_backend == "ascend"
+            )
+            if is_create_store:
+                from mf_adapter import create_config_store
```
Comment (ShangmingCai) [inline python/sglang/srt/managers/tokenizer_manager.py]: Should this line be put in the try-except as well? Maybe add a `ImportError` for `mf_adapter`?

File: python/sglang/srt/disaggregation/utils.py [hunk 20]
```diff
@@ -94,8 +95,12 @@ def __init__(
         custom_mem_pool: torch.cuda.MemPool = None,
     ):
         self.custom_mem_pool = custom_mem_pool
-        device = "cuda" if self.custom_mem_pool else "cpu"
-
+        device = "cpu"
+        if _is_npu:
+            # We need set output tokens to ascend device in order to transfer by D2D channel.
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/utils.py]: Sure. Thx for the explanation. However, consider optimizing this comment to avoid using 'we'? A better version would be "# For ascend backend, output tokens are placed in the NPU and will be transferred by D2D channel."

File: python/sglang/srt/disaggregation/utils.py [hunk 21]
```diff
@@ -15,8 +15,9 @@
 import torch
 import torch.distributed as dist
 
-from sglang.srt.utils import get_ip
+from sglang.srt.utils import get_ip, is_npu
 
+_is_npu = is_npu()
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/utils.py]: Is this flag gonna be used in many places? Is there a reason why we make it a global variable?

File: python/sglang/srt/disaggregation/ascend/transfer_engine.py [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+
+    def __init__(
+        self, hostname: str, npu_id: int, disaggregation_mode: DisaggregationMode
+    ):
+        try:
+            from mf_adapter import TransferEngine
+        except ImportError as e:
+            raise ImportError(
+                "Please install AscendTransferEngine, run SGLang with AscendTransferEngine"
+            ) from e
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/ascend/transfer_engine.py]: I think this import error info is not helpful. At least for me, I still don't know how to install **_mf_adapter_**.

**Review by ShangmingCai — COMMENTED** [review]
LGTM. Just some minor comments, then I think this PR should be good to merge.

One last question, IIUC, with this PR, we should be able to run PD with NPU through:
 - Use mooncake backend with ENABLE_ASCEND_TRANSFER_WITH_MOONCAKE=true
 - Use ascend backend with mf_adapter and ASCEND_MF_STORE_URL being set

If that is correct, consider adding this message in _sglang/docs/backend/pd_disaggregation.md_.

**Comment by zuochunwei** [comment]
does the PR support transferring kv-cache between npu(prefill) and gpu(decode)?

**Review by ShangmingCai — APPROVED** [review]
LGTM. Let me trigger the CI.

**Comment by ShangmingCai** [comment]
@zuochunwei Nope, hetero-device support might require kvstore-based method.

**Comment by ping1jing2** [comment]
@zhyncs please review and merge this PR, thank you.
